raiden/transfer/mediated_transfer/mediator.py

```diff
@@ -1149,12 +1187,58 @@ def handle_lock_expired(
     return TransitionResult(mediator_state, events)
 
 
+def handle_node_change_network_state(
+        mediator_state: MediatorTransferState,
+        state_change: ActionChangeNodeNetworkState,
+        channelidentifiers_to_channels: typing.ChannelMap,
+        pseudo_random_generator: random.Random,
+        block_number: typing.BlockNumber,
+) -> TransitionResult:
+    """ If a certain node comes online:
+    1. Check if a channel exists with that node
+    2. Check that this channel is a route, check if the route is valid.
+    3. Check that the transfer was stuck because there was no route available.
+    4. Send the transfer again to this now-available route.
+    """
+    if state_change.network_state != NODE_NETWORK_REACHABLE:
+        return TransitionResult(mediator_state, list())
+
+    # Get route whose node_address matches the newly reachable node
+    try:
+        route = next(route for route in mediator_state.routes
+                     if route.node_address == state_change.node_address)
+    except StopIteration:
+        # No routes were found.
+        return TransitionResult(mediator_state, list())
+
+    payer_channel = channelidentifiers_to_channels.get(mediator_state.payer_channel_identifier)
+    payee_channel = channelidentifiers_to_channels.get(route.channel_identifier)
+
+    payee_channel_open = channel.get_status(payee_channel) == CHANNEL_STATE_OPENED
+    if not payee_channel or not payee_channel_open:
+        return TransitionResult(mediator_state, list())
+
+    return mediate_transfer(
```

Comment:
This is wrong. It will mediate a transfer every time a node changes network status.

The correct behavior is to *only* send a new mediated transfer, iff the *last* received mediated transfer has not been mediated yet. (point 4 in the docstring)

Reply:
I used the approach that we discussed offline. 
When we fail to mediate the transfer and/or refund it, i store this `unrouted_transfer` for later so that it can be used. 
In the happy case scenario, the mediator_state will always have it set to None if we manage to route the transfer to an available node. 
The unhappy case happens when we can't forward the transfer... that's when we store it and when one of the nodes that is part of the route becomes available, we check if we have an `unrouted_transfer` and route it to the newly available node.